File: regression/_shared/log_query.py
"""Attested-log query DSL for TN-native assertions.

Whenever a regression test is checking TN's own protocol output
(envelopes in an attested log), the assertion should go through
`LogQuery.assert_contains(...)` — NOT a bare equality check, NOT
plumbed through generic `assert_named`.

Why a dedicated DSL: log queries have a specific failure-output shape
that's much more useful than "expected list with item, got list."
On miss we want:

- The named predicate spelled out.
- A summary of what IS in the log (event_type counts, latest sequence).
- A "closest match" envelope if one exists (same event_type, different fields).
- A pointer to where the test thinks the envelope should have come from.

The query operates against the raw attested log file (NDJSON). It does
NOT depend on `tn` being initialized — every regression test should be
able to read any log by path, regardless of whether the runtime that
produced it is alive.

See `_shared/README.md` for the contract + Style-1 example.
"""
from __future__ import annotations

import dataclasses
import json
from collections import Counter
from pathlib import Path
from typing import Any, Iterator

from regression._shared.assertions import (
    AssertionRecord,
    NamedAssertionError,
    _record,
    _resolve_silo,
    _resolve_test,
)


# ---------------------------------------------------------------------------
# Public surface
# ---------------------------------------------------------------------------


@dataclasses.dataclass
class Envelope:
    """One row from an attested NDJSON log. Holds the raw dict + a
    convenience accessor for top-level fields."""

    raw: dict[str, Any]

    def __getitem__(self, key: str) -> Any:
        return self.raw.get(key)

    def get(self, key: str, default: Any = None) -> Any:
        return self.raw.get(key, default)

    @property
    def event_type(self) -> str:
        return str(self.raw.get("event_type", ""))

    @property
    def sequence(self) -> int:
        s = self.raw.get("sequence")
        return int(s) if isinstance(s, int) else 0

    @property
    def row_hash(self) -> str:
        return str(self.raw.get("row_hash", ""))


class LogQuery:
    """Query against one or more attested NDJSON log files.

    Construct with:
        - `ceremony_path=<path>` — yaml file; reads its resolved
          `logs.path` plus the admin log if separate. Use this when
          you have a `LoadedConfig`-style ceremony in scope.
        - `log_paths=[<path>, <path>, …]` — explicit file list. Use
          this when the test knows exactly which logs to inspect.

    Methods:
        - `envelopes()` — iterator over all envelopes in chrono order.
        - `find_all(where=...)` — list of matching envelopes.
        - `find_one(where=...)` — first match or None.
        - `assert_contains(name=..., where=..., on_miss=...)` — named
          assertion; fails with a structured report on miss.
    """

    def __init__(
        self,
        *,
        ceremony_path: Path | str | None = None,
        log_paths: list[Path | str] | None = None,
    ) -> None:
        if ceremony_path and log_paths:
            raise ValueError(
                "LogQuery: pass either ceremony_path OR log_paths, not both"
            )
        if not ceremony_path and not log_paths:
            raise ValueError(
                "LogQuery: must pass ceremony_path or log_paths"
            )

        if ceremony_path:
            self._log_paths = _resolve_ceremony_logs(Path(ceremony_path))
        else:
            assert log_paths is not None  # narrow for type-checkers
            self._log_paths = [Path(p) for p in log_paths]

    # -- read paths ---------------------------------------------------------

    def envelopes(self) -> Iterator[Envelope]:
        """Yield every envelope from every log path in chronological
        order. Skips malformed lines silently — a regression test on
        malformed data should assert on the malformed-line case
        explicitly, not rely on the iterator to crash."""
        rows: list[tuple[str, dict[str, Any]]] = []
        for path in self._log_paths:
            if not path.exists():
                continue
            for line in path.read_text(encoding="utf-8").splitlines():
                line = line.strip()
                if not line:
                    continue
                try:
                    env = json.loads(line)
                except json.JSONDecodeError:
                    continue
                if not isinstance(env, dict):
                    continue
                ts = str(env.get("timestamp", ""))
                rows.append((ts, env))
        rows.sort(key=lambda r: r[0])
        for _, env in rows:
            yield Envelope(env)

    # -- find -------------------------------------------------------------

    def find_all(self, *, where: dict[str, Any]) -> list[Envelope]:
        """All envelopes matching every key/value in `where`. Top-level
        envelope fields only (no plaintext-group field matching yet —
        decryption is the runtime's job, not the assertion's)."""
        out: list[Envelope] = []
        for env in self.envelopes():
            if _matches(env.raw, where):
                out.append(env)
        return out

    def find_one(self, *, where: dict[str, Any]) -> Envelope | None:
        for env in self.envelopes():
            if _matches(env.raw, where):
                return env
        return None

    def event_type_counts(self) -> dict[str, int]:
        c: Counter[str] = Counter()
        for env in self.envelopes():
            c[env.event_type] += 1
        return dict(c)

    # -- named assertion --------------------------------------------------

    def assert_contains(
        self,
        *,
        name: str,
        where: dict[str, Any],
        on_miss: str | None = None,
    ) -> Envelope:
        """Assert at least one envelope matches `where`. Returns the
        first match for further inspection (caller can chain
        additional checks).

        On miss:
            - Prints the named predicate.
            - Prints the event_type histogram for the log.
            - Prints the "closest match" (same event_type, different
              fields) if one exists — usually the right next thing
              to look at.
            - Prints the operator-supplied `on_miss` pointer.
        """
        matches = self.find_all(where=where)
        silo = _resolve_silo()
        test = _resolve_test()

        if matches:
            _record(
                AssertionRecord(
                    name=name,
                    style="log-query",
                    passed=True,
                    expected=f"at least 1 envelope where {_pp(where)}",
                    observed=f"{len(matches)} match(es)",
                    on_miss=on_miss or "",
                    silo=silo,
                    test=test,
                )
            )
            return matches[0]

        # MISS — build the structured failure report.
        et = where.get("event_type")
        counts = self.event_type_counts()
        closest = _find_closest(self.envelopes(), where) if et else None
        miss_pointer = on_miss or "(no pointer supplied — add `on_miss=` to the assertion)"

        observed_repr = (
            f"no envelope matched; event_type counts in log = "
            f"{_pp(counts)}; closest match = {_pp(closest.raw) if closest else 'none'}"
        )

        _record(
            AssertionRecord(
                name=name,
                style="log-query",
                passed=False,
                expected=f"at least 1 envelope where {_pp(where)}",
                observed=observed_repr,
                on_miss=miss_pointer,
                silo=silo,
                test=test,
            )
        )

        msg_lines = [
            f"ASSERTION FAILED: {name}",
            f"  silo: {silo}",
            f"  test: {test}",
            f"  style: log-query",
            f"  predicate: {_pp(where)}",
            f"  observed in log:",
            f"    paths: {[str(p) for p in self._log_paths]}",
            f"    total event_types: {_pp(counts)}",
        ]
        if closest is not None:
            msg_lines.append(f"  closest match (same event_type): {_pp(closest.raw)}")
        else:
            msg_lines.append(f"  closest match: <none — no envelope had event_type={et!r}>")
        msg_lines.append(f"  look at: {miss_pointer}")
        raise NamedAssertionError("\n".join(msg_lines))


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def _matches(env: dict[str, Any], where: dict[str, Any]) -> bool:
    """Predicate: every (k, v) in `where` must equal env[k]. Missing
    keys count as a miss; nested matching is NOT supported (assertion
    on group-plaintext requires decryption, out of scope for
    log-level queries)."""
    for k, v in where.items():
        if env.get(k) != v:
            return False
    return True


def _find_closest(
    envs: Iterator[Envelope], where: dict[str, Any]
) -> Envelope | None:
    """Best-effort: an envelope with matching event_type but
    differing on at least one other key. Helpful in failure output
    so the maintainer can see "you got a tn.recipient.added but for
    a different DID."""
    et = where.get("event_type")
    if not et:
        return None
    for env in envs:
        if env.event_type == et:
            return env
    return None


def _pp(value: Any) -> str:
    """Compact repr suitable for failure messages — never None,
    never raises."""
    if value is None:
        return "None"
    try:
        return json.dumps(value, default=str, sort_keys=True)
    except (TypeError, ValueError):
        return repr(value)


def _resolve_ceremony_logs(yaml_path: Path) -> list[Path]:
    """Resolve a ceremony yaml to its log file list (main + admin if
    separate). Hand-rolled; does NOT depend on `tn.config.load()`
    because we want the regression suite to be inspectable even when
    the runtime is broken.

    Convention: read just enough yaml to find `logs.path` and
    optionally `ceremony.admin_log_location`. PyYAML is a hard dep
    of tn-protocol so it's always available.
    """
    import yaml

    try:
        doc = yaml.safe_load(yaml_path.read_text(encoding="utf-8")) or {}
    except (OSError, yaml.YAMLError):
        return []
    if not isinstance(doc, dict):
        return []

    base = yaml_path.parent
    paths: list[Path] = []

    logs = doc.get("logs") or {}
    main = logs.get("path") if isinstance(logs, dict) else None
    if isinstance(main, str):
        paths.append((base / main).resolve())

    cer = doc.get("ceremony") or {}
    admin = cer.get("admin_log_location") if isinstance(cer, dict) else None
    if isinstance(admin, str) and admin not in ("main_log", "") and "{" not in admin:
        paths.append((base / admin).resolve())

    return paths

File: regression/_shared/log_query.ts
// Attested-log query DSL for TN-native assertions (TS side).
//
// Mirrors `regression/_shared/log_query.py` — same predicates, same
// failure-output shape. Whenever a regression test is checking TN's own
// protocol output (envelopes in an attested log), the assertion goes
// through `LogQuery.assertContains(...)` — NOT bare equality, NOT a
// generic `assertNamed`.
//
// See `_shared/README.md` for the contract + Style-1 example.

import { existsSync, readFileSync } from "node:fs";
import { dirname, resolve as resolvePath } from "node:path";
import { parse as parseYaml } from "yaml";

import {
  AssertionRecord,
  NamedAssertionError,
  _formatFailureForLogQuery,
  _recordForLogQuery,
  _resolveSiloForLogQuery,
  _resolveTestForLogQuery,
} from "./assertions.js";

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

export interface RawEnvelope {
  [key: string]: unknown;
}

export class Envelope {
  constructor(public readonly raw: RawEnvelope) {}

  get eventType(): string {
    return String(this.raw["event_type"] ?? "");
  }

  get sequence(): number {
    const s = this.raw["sequence"];
    return typeof s === "number" ? s : 0;
  }

  get rowHash(): string {
    return String(this.raw["row_hash"] ?? "");
  }

  get(key: string): unknown {
    return this.raw[key];
  }
}

export type Predicate = Record<string, unknown>;

export interface LogQueryArgs {
  /** Path to the ceremony yaml. Resolves logs.path + admin_log_location. */
  ceremonyPath?: string;
  /** Explicit list of log file paths. Mutually exclusive with ceremonyPath. */
  logPaths?: string[];
}

export interface AssertContainsArgs {
  name: string;
  where: Predicate;
  onMiss?: string;
}

// ---------------------------------------------------------------------------
// LogQuery
// ---------------------------------------------------------------------------

export class LogQuery {
  private readonly logPaths: string[];

  constructor(args: LogQueryArgs) {
    if (args.ceremonyPath && args.logPaths) {
      throw new Error("LogQuery: pass either ceremonyPath OR logPaths, not both");
    }
    if (!args.ceremonyPath && !args.logPaths) {
      throw new Error("LogQuery: must pass ceremonyPath or logPaths");
    }

    this.logPaths = args.ceremonyPath
      ? resolveCeremonyLogs(args.ceremonyPath)
      : args.logPaths!.map((p) => resolvePath(p));
  }

  /** All envelopes from every log path, chronological order, malformed
   *  lines silently skipped. */
  envelopes(): Envelope[] {
    type Row = { ts: string; env: RawEnvelope };
    const rows: Row[] = [];
    for (const path of this.logPaths) {
      if (!existsSync(path)) continue;
      const text = readFileSync(path, "utf-8");
      for (const rawLine of text.split(/\r?\n/)) {
        const line = rawLine.trim();
        if (!line) continue;
        let parsed: unknown;
        try {
          parsed = JSON.parse(line);
        } catch {
          continue;
        }
        if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) continue;
        const env = parsed as RawEnvelope;
        const ts = String(env["timestamp"] ?? "");
        rows.push({ ts, env });
      }
    }
    rows.sort((a, b) => (a.ts < b.ts ? -1 : a.ts > b.ts ? 1 : 0));
    return rows.map((r) => new Envelope(r.env));
  }

  findAll(args: { where: Predicate }): Envelope[] {
    return this.envelopes().filter((env) => matches(env.raw, args.where));
  }

  findOne(args: { where: Predicate }): Envelope | null {
    return this.findAll(args)[0] ?? null;
  }

  eventTypeCounts(): Record<string, number> {
    const out: Record<string, number> = {};
    for (const env of this.envelopes()) {
      const et = env.eventType;
      out[et] = (out[et] ?? 0) + 1;
    }
    return out;
  }

  /**
   * Assert at least one envelope matches `where`. Returns the first
   * match for chained inspection. On miss, throws NamedAssertionError
   * with the same structured shape as the Python equivalent.
   */
  assertContains(args: AssertContainsArgs): Envelope {
    const matchesList = this.findAll({ where: args.where });
    const silo = _resolveSiloForLogQuery();
    const test = _resolveTestForLogQuery();

    if (matchesList.length > 0) {
      _recordForLogQuery({
        name: args.name,
        style: "log-query",
        passed: true,
        expected: `at least 1 envelope where ${pp(args.where)}`,
        observed: `${matchesList.length} match(es)`,
        on_miss: args.onMiss ?? "",
        silo,
        test,
      });
      return matchesList[0]!;
    }

    // MISS — build structured failure
    const counts = this.eventTypeCounts();
    const et = args.where["event_type"];
    const closest = et !== undefined ? findClosest(this.envelopes(), args.where) : null;
    const missPointer =
      args.onMiss ?? "(no pointer supplied — add `onMiss` to the assertion)";

    const observedRepr =
      `no envelope matched; event_type counts in log = ${pp(counts)}; ` +
      `closest match = ${closest ? pp(closest.raw) : "none"}`;

    _recordForLogQuery({
      name: args.name,
      style: "log-query",
      passed: false,
      expected: `at least 1 envelope where ${pp(args.where)}`,
      observed: observedRepr,
      on_miss: missPointer,
      silo,
      test,
    });

    const lines = [
      `ASSERTION FAILED: ${args.name}`,
      `  silo: ${silo}`,
      `  test: ${test}`,
      `  style: log-query`,
      `  predicate: ${pp(args.where)}`,
      `  observed in log:`,
      `    paths: ${pp(this.logPaths)}`,
      `    total event_types: ${pp(counts)}`,
    ];
    if (closest) {
      lines.push(`  closest match (same event_type): ${pp(closest.raw)}`);
    } else {
      lines.push(`  closest match: <none — no envelope had event_type=${pp(et)}>`);
    }
    lines.push(`  look at: ${missPointer}`);

    // Use the formatter to keep parity (covers the rare future case
    // where formatFailure changes shape and we want both styles to
    // track in lockstep).
    void _formatFailureForLogQuery;

    throw new NamedAssertionError(lines.join("\n"));
  }
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function matches(env: RawEnvelope, where: Predicate): boolean {
  for (const [k, v] of Object.entries(where)) {
    if (env[k] !== v) return false;
  }
  return true;
}

function findClosest(envs: Envelope[], where: Predicate): Envelope | null {
  const et = where["event_type"];
  if (!et) return null;
  return envs.find((e) => e.eventType === et) ?? null;
}

function pp(v: unknown): string {
  if (v === null || v === undefined) return "None";
  try {
    return JSON.stringify(v);
  } catch {
    return String(v);
  }
}

/**
 * Resolve a ceremony yaml to its log file list (main + admin if
 * separate). **Behavior contract: must match Python's
 * `_resolve_ceremony_logs` in `regression/_shared/log_query.py`
 * byte-for-byte for the same input.** Both sides use a real yaml
 * parser (PyYAML on Python, the `yaml` npm package on TS) so the
 * regression suite asserts the same files on both runtimes regardless
 * of anchors, block scalars, multi-doc files, or quoted keys.
 *
 * Algorithm (mirrors Python):
 *   1. Parse the yaml. Non-dict / parse-error → return [].
 *   2. `doc.logs.path` (if `logs` is a dict + `path` is a string)
 *      → append, resolved against yaml's parent dir.
 *   3. `doc.ceremony.admin_log_location` (if `ceremony` is a dict
 *      + value is a string AND not `"main_log"` AND not empty AND
 *      contains no `{` template tokens) → append, resolved against
 *      yaml's parent.
 *   4. Return.
 *
 * The regression suite intentionally does NOT depend on the SDK's
 * `loadConfig` — when the SDK config loader is broken, the regression
 * suite needs to still work to surface the bug.
 */
function resolveCeremonyLogs(yamlPath: string): string[] {
  if (!existsSync(yamlPath)) return [];

  let doc: unknown;
  try {
    doc = parseYaml(readFileSync(yamlPath, "utf-8"));
  } catch {
    return [];
  }
  if (doc === null || doc === undefined) return [];
  if (typeof doc !== "object" || Array.isArray(doc)) return [];

  const d = doc as Record<string, unknown>;
  const base = dirname(yamlPath);
  const out: string[] = [];

  // 1. logs.path
  const logsBlock = d["logs"];
  if (logsBlock !== null && typeof logsBlock === "object" && !Array.isArray(logsBlock)) {
    const main = (logsBlock as Record<string, unknown>)["path"];
    if (typeof main === "string") {
      out.push(resolvePath(base, main));
    }
  }

  // 2. ceremony.admin_log_location
  const cerBlock = d["ceremony"];
  if (cerBlock !== null && typeof cerBlock === "object" && !Array.isArray(cerBlock)) {
    const admin = (cerBlock as Record<string, unknown>)["admin_log_location"];
    if (
      typeof admin === "string" &&
      admin !== "main_log" &&
      admin !== "" &&
      !admin.includes("{")
    ) {
      out.push(resolvePath(base, admin));
    }
  }

  // Keep AssertionRecord referenced — the import above is for type
  // consumers but we don't currently use it inside this function.
  void ({} as AssertionRecord);

  return out;
}
